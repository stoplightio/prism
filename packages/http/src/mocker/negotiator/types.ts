import { IHttpOperationConfig } from '@stoplight/prism-http';
import { INodeExample, INodeExternalExample, ISchema } from '@stoplight/types';

export interface IHttpNegotiationResult {
  code: string;
  mediaType: string;
  example?: INodeExample | INodeExternalExample;
  schema?: ISchema;
}

export type NegotiationOptions = IHttpOperationConfig;

export type NegotiatePartialOptions = {
  code: string;
  dynamic: boolean;
  exampleKey?: string;
};
