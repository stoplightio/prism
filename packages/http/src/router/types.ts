import { IHttpOperationEx } from '../types';

export type Nullable<T> = T | null;

export interface IMatch {
  resource: IHttpOperationEx;
  pathMatch: MatchType;
  methodMatch: MatchType;
  serverMatch?: Nullable<MatchType>;
}

export enum MatchType {
  CONCRETE = 'concrete',
  TEMPLATED = 'templated',
  NOMATCH = 'no-match',
}
