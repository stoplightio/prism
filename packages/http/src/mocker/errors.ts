import { ProblemJson } from '../types';

export const UNPROCESSABLE_ENTITY: Omit<ProblemJson, 'detail'> = {
  type: 'UNPROCESSABLE_ENTITY',
  title: 'Invalid request body payload',
  status: 422,
};

export const NOT_ACCEPTABLE: Omit<ProblemJson, 'detail'> = {
  type: 'NOT_ACCEPTABLE',
  title: 'The server cannot produce a representation for your accept header',
  status: 406,
};

export const NOT_FOUND: Omit<ProblemJson, 'detail'> = {
  type: 'NOT_FOUND',
  title: 'The server cannot find the requested content',
  status: 404,
};

export const UNAUTHORIZED: Omit<ProblemJson, 'detail'> = {
  type: 'UNAUTHORIZED',
  title: 'Invalid security scheme used',
  status: 401,
};

export const VIOLATIONS: Omit<ProblemJson, 'detail'> = {
  type: 'VIOLATIONS',
  title: 'Request/Response not valid',
  status: 500,
};
