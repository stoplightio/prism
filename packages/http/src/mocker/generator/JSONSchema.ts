// @ts-ignore
import * as jsf from '@stoplight/json-schema-faker';
import { cloneDeep } from 'lodash';

jsf.option({
  failOnInvalidTypes: false,
  failOnInvalidFormat: false,
  alwaysFakeOptionals: true,
  optionalsProbability: 1,
  fixedProbabilities: true,
  ignoreMissingRefs: true,
  useExamplesValue: true,
  useDefaultValue: true,
  maxItems: 20,
  maxLength: 100,
});

export async function generate(source: unknown): Promise<unknown> {
  return jsf.resolve(cloneDeep(source));
}
