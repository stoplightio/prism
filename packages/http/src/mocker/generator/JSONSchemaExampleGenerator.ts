// @ts-ignore
import * as jsf from '@stoplight/json-schema-faker';
import { cloneDeep } from 'lodash';
import { IExampleGenerator } from './IExampleGenerator';

jsf.option({
  failOnInvalidTypes: false,
  failOnInvalidFormat: false,
  alwaysFakeOptionals: true,
  optionalsProbability: 1,
  fixedProbabilities: true,
  ignoreMissingRefs: true,
  useExamplesValue: true,
  useDefaultValue: true,
  maxItems: 20,
  maxLength: 100,
});

export class JSONSchemaExampleGenerator implements IExampleGenerator {
  public async generate(schema: unknown, mediaType: string): Promise<string> {
    const example = await jsf.resolve(cloneDeep(schema));
    return this.transform(mediaType, example);
  }

  private transform(mediaType: string, input: object): string {
    switch (mediaType) {
      case 'application/json':
        return JSON.stringify(input);

      default:
        throw new Error(`Unknown media type '${mediaType}'`);
    }
  }
}
