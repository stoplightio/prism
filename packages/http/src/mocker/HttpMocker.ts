import { IMocker, IMockerOpts } from '@stoplight/prism-core';
import { Dictionary, IHttpHeaderParam, IHttpOperation, INodeExample, INodeExternalExample } from '@stoplight/types';

import * as caseless from 'caseless';
import { fromPairs, isEmpty, isObject, keyBy, mapValues, toPairs } from 'lodash';
import {
  IHttpConfig,
  IHttpOperationConfig,
  IHttpRequest,
  IHttpResponse,
  PayloadGenerator,
  ProblemJsonError,
} from '../types';
import { UNPROCESSABLE_ENTITY } from './errors';
import helpers from './negotiator/NegotiatorHelpers';
import { IHttpNegotiationResult } from './negotiator/types';

export class HttpMocker implements IMocker<IHttpOperation, IHttpRequest, IHttpConfig, IHttpResponse> {
  constructor(private _exampleGenerator: PayloadGenerator) {}

  public async mock({
    resource,
    input,
    config,
  }: Partial<IMockerOpts<IHttpOperation, IHttpRequest, IHttpConfig>>): Promise<IHttpResponse> {
    // pre-requirements check
    if (!resource) {
      throw new Error('Resource is not defined');
    }

    if (!input) {
      throw new Error('Http request is not defined');
    }

    // setting default values
    const inputMediaType = input.data.headers && caseless(input.data.headers).get('content-type');
    config = config || { mock: false };
    const mockConfig: IHttpOperationConfig =
      config.mock === false ? { dynamic: false } : Object.assign({}, config.mock);

    if (!mockConfig.mediaType && typeof inputMediaType === 'string') {
      mockConfig.mediaType = inputMediaType;
    }

    // looking up proper example
    let negotiationResult: IHttpNegotiationResult;
    if (input.validations.input.length > 0) {
      try {
        negotiationResult = helpers.negotiateOptionsForInvalidRequest(resource.responses);
      } catch (error) {
        throw ProblemJsonError.fromTemplate(
          UNPROCESSABLE_ENTITY,
          `Your request body is not valid: ${JSON.stringify(input.validations.input)}`,
        );
      }
    } else {
      negotiationResult = helpers.negotiateOptionsForValidRequest(resource, mockConfig);
    }

    const [body, mockedHeaders] = await Promise.all([
      computeBody(negotiationResult, this._exampleGenerator),
      computeMockedHeaders(negotiationResult.headers, this._exampleGenerator),
    ]);

    return {
      statusCode: parseInt(negotiationResult.code),
      headers: {
        ...mockedHeaders,
        'Content-type': negotiationResult.mediaType,
      },
      body,
    };
  }
}

function isINodeExample(nodeExample: INodeExample | INodeExternalExample | undefined): nodeExample is INodeExample {
  return !!nodeExample && 'value' in nodeExample;
}

function computeMockedHeaders(headers: IHttpHeaderParam[], ex: PayloadGenerator): Promise<Dictionary<string>> {
  const headerWithPromiseValues = mapValues(keyBy(headers, h => h.name), async header => {
    if (header.content) {
      if (header.content.examples.length > 0) {
        const example = header.content.examples[0];
        if (isINodeExample(example)) {
          return example.value;
        }
      }
      if (header.content.schema) {
        const example = await ex(header.content.schema);
        if (!(isObject(example) && isEmpty(example))) return example;
      }
    }
    return '';
  });

  return resolvePromiseInProps(headerWithPromiseValues);
}

async function computeBody(
  negotiationResult: Pick<IHttpNegotiationResult, 'schema' | 'mediaType' | 'bodyExample'>,
  ex: PayloadGenerator,
) {
  if (isINodeExample(negotiationResult.bodyExample) && negotiationResult.bodyExample.value !== undefined) {
    return negotiationResult.bodyExample.value;
  } else if (negotiationResult.schema) {
    return ex(negotiationResult.schema);
  }
  return undefined;
}

async function resolvePromiseInProps(val: Dictionary<Promise<string>>): Promise<Dictionary<string>> {
  const promisePair = await Promise.all(toPairs(val).map(v => Promise.all(v)));
  return fromPairs(promisePair);
}
