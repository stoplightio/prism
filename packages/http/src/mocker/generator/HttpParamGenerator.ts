import { IHttpContent, IHttpParam, INodeExample, INodeExternalExample } from '@stoplight/types';
import * as O from 'fp-ts/lib/Option';
import { pipe } from 'fp-ts/lib/pipeable';
import { JSONSchema } from '../../types';
import { generate as generateDynamicExample } from './JSONSchema';

export function improveSchema(schema: JSONSchema): JSONSchema {
  const newSchema = { ...schema };

  if (newSchema.type === 'integer' || newSchema.type === 'number') {
    if (!newSchema.minimum) {
      newSchema.minimum = 1;
    }

    if (!newSchema.maximum) {
      newSchema.maximum = 1000;
    }
  }

  if (newSchema.type === 'string') {
    if (!newSchema.format && !newSchema.enum && !newSchema.pattern) {
      newSchema['x-faker'] = 'lorem.word';
    }
  }

  if (newSchema.type === 'object') {
    if (newSchema.properties) {
      newSchema.properties = Object.entries(newSchema.properties).reduce((r, [k, v]) => {
        r[k] = improveSchema(v);
        return r;
      }, {});
    }
  }

  if (newSchema.type === 'array') {
    if (typeof newSchema.items === 'object') {
      newSchema.items = improveSchema(newSchema.items);
    }
  }

  return newSchema;
}

function pickStaticExample(examples: O.Option<Array<INodeExample | INodeExternalExample>>): O.Option<unknown> {
  return pipe(
    examples,
    O.mapNullable(exs => exs[Math.floor(Math.random() * exs.length)]),
    O.mapNullable(example => (example as INodeExample).value)
  );
}

export function generate(param: IHttpParam | IHttpContent): O.Option<unknown> {
  return pipe(
    O.fromNullable(param.examples),
    pickStaticExample,
    O.alt(() =>
      pipe(
        O.fromNullable(param.schema),
        O.map(improveSchema),
        O.chain(schema => O.fromEither(generateDynamicExample(schema)))
      )
    )
  );
}
