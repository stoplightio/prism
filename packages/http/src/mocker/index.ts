import { IPrismComponents, IPrismInput } from '@stoplight/prism-core';
import {
  DiagnosticSeverity,
  IHttpHeaderParam,
  IHttpOperation,
  INodeExample,
  IMediaTypeContent,
} from '@stoplight/types';

import * as caseless from 'caseless';
import * as E from 'fp-ts/lib/Either';
import * as Record from 'fp-ts/lib/Record';
import { pipe } from 'fp-ts/lib/pipeable';
import * as R from 'fp-ts/lib/Reader';
import * as O from 'fp-ts/lib/Option';
import * as RE from 'fp-ts/lib/ReaderEither';
import { map } from 'fp-ts/lib/Array';
import { isNumber, isString, keyBy, mapValues, groupBy, get } from 'lodash';
import { Logger } from 'pino';
import * as typeIs from 'type-is';
import {
  ContentExample,
  IHttpOperationConfig,
  IHttpRequest,
  IHttpResponse,
  IMockHttpConfig,
  PayloadGenerator,
  ProblemJsonError,
} from '../types';
import withLogger from '../withLogger';
import { UNAUTHORIZED, UNPROCESSABLE_ENTITY } from './errors';
import { generate, generateStatic } from './generator/JSONSchema';
import helpers from './negotiator/NegotiatorHelpers';
import { IHttpNegotiationResult } from './negotiator/types';
import { runCallback } from './callback/callbacks';
import {
  decodeUriEntities,
  deserializeFormBody,
  findContentByMediaTypeOrFirst,
  splitUriParams,
} from '../validator/validators/body';
import { sequenceT } from 'fp-ts/lib/Apply';

const eitherRecordSequence = Record.sequence(E.either);
const eitherSequence = sequenceT(E.either);

const mock: IPrismComponents<IHttpOperation, IHttpRequest, IHttpResponse, IMockHttpConfig>['mock'] = ({
  resource,
  input,
  config,
}) => {
  const payloadGenerator: PayloadGenerator = config.dynamic ? generate : generateStatic;

  return pipe(
    withLogger(logger => {
      // setting default values
      const acceptMediaType = input.data.headers && caseless(input.data.headers).get('accept');
      if (!config.mediaTypes && acceptMediaType) {
        logger.info(`Request contains an accept header: ${acceptMediaType}`);
        config.mediaTypes = acceptMediaType.split(',');
      }

      return config;
    }),
    R.chain(mockConfig => negotiateResponse(mockConfig, input, resource)),
    R.chain(result => assembleResponse(result, payloadGenerator)),
    R.chain(response =>
      /*  Note: This is now just logging the errors without propagating them back. This might be moved as a first
          level concept in Prism.
      */
      logger =>
        pipe(
          response,
          E.map(response => runCallbacks({ resource, request: input.data, response })(logger)),
          E.chain(() => response)
        )
    )
  );
};

function runCallbacks({
  resource,
  request,
  response,
}: {
  resource: IHttpOperation;
  request: IHttpRequest;
  response: IHttpResponse;
}) {
  return withLogger(logger =>
    pipe(
      O.fromNullable(resource.callbacks),
      O.map(callbacks =>
        pipe(
          callbacks,
          map(callback =>
            runCallback({ callback, request: parseBodyIfUrlEncoded(request, resource), response })(logger)()
          )
        )
      )
    )
  );
}

/*
  This function should not be here at all, but unfortunately due to some limitations of the Monad we're using (Either)
  we cannot carry parsed informations in case of an error â€” which is what we do need instead.
*/
function parseBodyIfUrlEncoded(request: IHttpRequest, resource: IHttpOperation) {
  const mediaType = caseless(request.headers || {}).get('content-type');
  if (!mediaType) return request;

  if (!typeIs.is(mediaType, ['application/x-www-form-urlencoded'])) return request;

  const specs = pipe(
    O.fromNullable(resource.request),
    O.mapNullable(request => request.body),
    O.mapNullable(body => body.contents),
    O.getOrElse(() => [] as IMediaTypeContent[])
  );

  const encodedUriParams = splitUriParams(request.body as string);

  if (specs.length < 1) {
    return Object.assign(request, { body: encodedUriParams });
  }

  const content = pipe(
    O.fromNullable(mediaType),
    O.chain(mediaType => findContentByMediaTypeOrFirst(specs, mediaType)),
    O.map(({ content }) => content),
    O.getOrElse(() => specs[0] || {})
  );

  const encodings = get(content, 'encodings', []);

  if (!content.schema) return Object.assign(request, { body: encodedUriParams });

  return Object.assign(request, {
    body: deserializeFormBody(content.schema, encodings, decodeUriEntities(encodedUriParams)),
  });
}

function handleInputValidation(input: IPrismInput<IHttpRequest>, resource: IHttpOperation) {
  const securityValidation = input.validations.find(validation => validation.code === 401);

  return pipe(
    withLogger(logger => logger.warn({ name: 'VALIDATOR' }, 'Request did not pass the validation rules')),
    R.chain(() =>
      pipe(
        helpers.negotiateOptionsForInvalidRequest(resource.responses, securityValidation ? ['401'] : ['422', '400']),
        RE.mapLeft(() =>
          securityValidation
            ? ProblemJsonError.fromTemplate(
                UNAUTHORIZED,
                '',
                securityValidation.tags && securityValidation.tags.length
                  ? {
                      headers: { 'WWW-Authenticate': securityValidation.tags.join(',') },
                    }
                  : undefined
              )
            : ProblemJsonError.fromTemplate(
                UNPROCESSABLE_ENTITY,
                'Your request is not valid and no HTTP validation response was found in the spec, so Prism is generating this error for you.',
                {
                  validation: input.validations.map(detail => ({
                    location: detail.path,
                    severity: DiagnosticSeverity[detail.severity],
                    code: detail.code,
                    message: detail.message,
                  })),
                }
              )
        )
      )
    )
  );
}

function negotiateResponse(
  mockConfig: IHttpOperationConfig,
  input: IPrismInput<IHttpRequest>,
  resource: IHttpOperation
) {
  const { [DiagnosticSeverity.Error]: errors, [DiagnosticSeverity.Warning]: warnings } = groupBy(
    input.validations,
    validation => validation.severity
  );

  if (errors) {
    return handleInputValidation(input, resource);
  } else {
    return pipe(
      withLogger(logger => {
        warnings && warnings.forEach(warn => logger.warn({ name: 'VALIDATOR' }, warn.message));
        return logger.success(
          { name: 'VALIDATOR' },
          'The request passed the validation rules. Looking for the best response'
        );
      }),
      R.chain(() => helpers.negotiateOptionsForValidRequest(resource, mockConfig))
    );
  }
}

function assembleResponse(
  result: E.Either<Error, IHttpNegotiationResult>,
  payloadGenerator: PayloadGenerator
): R.Reader<Logger, E.Either<Error, IHttpResponse>> {
  return logger =>
    pipe(
      result,
      E.chain(negotiationResult =>
        pipe(
          eitherSequence(
            computeBody(negotiationResult, payloadGenerator),
            computeMockedHeaders(negotiationResult.headers || [], payloadGenerator)
          ),
          E.map(([mockedBody, mockedHeaders]) => {
            const response: IHttpResponse = {
              statusCode: parseInt(negotiationResult.code),
              headers: {
                ...mockedHeaders,
                ...(negotiationResult.mediaType && { 'Content-type': negotiationResult.mediaType }),
              },
              body: mockedBody,
            };

            logger.success(`Responding with the requested status code ${response.statusCode}`);

            return response;
          })
        )
      )
    );
}

function isINodeExample(nodeExample: ContentExample | undefined): nodeExample is INodeExample {
  return !!nodeExample && 'value' in nodeExample;
}

function computeMockedHeaders(headers: IHttpHeaderParam[], payloadGenerator: PayloadGenerator) {
  return eitherRecordSequence(
    mapValues(
      keyBy(headers, h => h.name),
      header => {
        if (header.schema) {
          if (header.examples && header.examples.length > 0) {
            const example = header.examples[0];
            if (isINodeExample(example)) {
              return E.right(example.value);
            }
          } else {
            return pipe(
              payloadGenerator(header.schema),
              E.map(example => {
                if (isNumber(example) || isString(example)) return example;
                return null;
              })
            );
          }
        }
        return E.right(null);
      }
    )
  );
}

function computeBody(
  negotiationResult: Pick<IHttpNegotiationResult, 'schema' | 'mediaType' | 'bodyExample'>,
  payloadGenerator: PayloadGenerator
): E.Either<Error, unknown> {
  if (isINodeExample(negotiationResult.bodyExample) && negotiationResult.bodyExample.value !== undefined) {
    return E.right(negotiationResult.bodyExample.value);
  } else if (negotiationResult.schema) {
    return payloadGenerator(negotiationResult.schema);
  }
  return E.right(undefined);
}

export default mock;
