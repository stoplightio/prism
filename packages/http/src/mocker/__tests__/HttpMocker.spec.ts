import { IHttpOperation, INodeExample } from '@stoplight/types';
import { flatMap } from 'lodash';
import { HttpMocker } from '../../mocker';
import { JSONSchemaExampleGenerator } from '../../mocker/generator/JSONSchemaExampleGenerator';
import helpers from '../negotiator/NegotiatorHelpers';

describe('HttpMocker', () => {
  const mockExampleGenerator = new JSONSchemaExampleGenerator();
  const httpMocker = new HttpMocker(mockExampleGenerator);

  afterEach(() => jest.restoreAllMocks());

  describe('mock()', () => {
    const mockSchema = {
      type: 'object',
      properties: {
        name: { type: 'string' },
        surname: { type: 'string', format: 'email' },
      },
      required: ['name', 'email'],
    };

    const mockResource: IHttpOperation = {
      id: 'id',
      method: 'get',
      path: '/test',
      servers: [],
      security: [],
      request: {
        headers: [],
        query: [],
        cookie: [],
        path: [],
      },
      responses: [
        {
          code: '200',
          headers: [],
          contents: [
            {
              mediaType: 'application/json',
              schema: mockSchema,
              examples: [
                {
                  key: 'preferred key',
                  value: 'hello',
                },
                {
                  key: 'test key',
                  value: 'test value',
                },
                {
                  key: 'test key2',
                  externalValue: 'http://example.org/examples/example1',
                },
              ],
              encodings: [],
            },
          ],
        },
      ],
    };

    const mockInput = {
      validations: {
        input: [],
      },
      data: {
        method: 'get' as const,
        url: {
          path: '/test',
          baseUrl: 'example.com',
        },
      },
    };

    it('fails when called with no resource', () => {
      return expect(
        httpMocker.mock({
          input: mockInput,
        }),
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it('fails when called with no input', () => {
      return expect(
        httpMocker.mock({
          resource: mockResource,
        }),
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    describe('with valid negotiator response', () => {
      it('returns an empty body when negotiator did not resolve to either example nor schema', () => {
        jest
          .spyOn(helpers, 'negotiateOptionsForValidRequest')
          .mockImplementation(() => ({ code: '202', mediaType: 'test' }));

        return expect(
          httpMocker.mock({
            resource: mockResource,
            input: mockInput,
          }),
        ).resolves.toHaveProperty('body', undefined);
      });

      it('returns static example', () => {
        jest.spyOn(helpers, 'negotiateOptionsForValidRequest').mockImplementation(() => ({
          code: '202',
          mediaType: 'test',
          example: mockResource.responses![0].contents![0].examples![0],
        }));

        return expect(
          httpMocker.mock({
            resource: mockResource,
            input: mockInput,
          }),
        ).resolves.toMatchSnapshot();
      });

      it('returns dynamic example', () => {
        jest.spyOn(helpers, 'negotiateOptionsForValidRequest').mockImplementation(() => ({
          code: '202',
          mediaType: 'test',
          schema: mockResource.responses![0].contents![0].schema,
        }));

        jest.spyOn(mockExampleGenerator, 'generate').mockResolvedValue('example value');

        return expect(
          httpMocker.mock({
            resource: mockResource,
            input: mockInput,
          }),
        ).resolves.toMatchSnapshot();
      });
    });

    describe('with invalid negotiator response', () => {
      it('returns static example', () => {
        jest.spyOn(helpers, 'negotiateOptionsForInvalidRequest').mockImplementation(() => ({
          code: '202',
          mediaType: 'test',
          example: mockResource.responses![0].contents![0].examples![0],
        }));

        return expect(
          httpMocker.mock({
            resource: mockResource,
            input: Object.assign({}, mockInput, { validations: { input: [{}] } }),
          }),
        ).resolves.toMatchSnapshot();
      });
    });

    describe('when example is of type INodeExternalExample', () => {
      it('generates a dynamic example', () => {
        jest.spyOn(helpers, 'negotiateOptionsForValidRequest').mockImplementation(() => ({
          code: '202',
          mediaType: 'test',
          example: mockResource.responses![0].contents![0].examples![1],
          schema: { type: 'string' },
        }));

        jest.spyOn(mockExampleGenerator, 'generate').mockResolvedValue('example value chelsea');

        return expect(
          httpMocker.mock({
            resource: mockResource,
            input: mockInput,
          }),
        ).resolves.toMatchSnapshot();
      });
    });

    describe('when an example is defined', () => {
      describe('and dynamic flag is true', () => {
        describe('should generate a dynamic response', () => {
          const generatedExample = JSON.stringify({ hello: 'world' });
          beforeAll(() => {
            jest.spyOn(mockExampleGenerator, 'generate').mockResolvedValue(generatedExample);
          });
          afterAll(() => {
            jest.restoreAllMocks();
          });

          it('the dynamic response should not be an example one', async () => {
            const response = await httpMocker.mock({
              input: mockInput,
              resource: mockResource,
              config: { mock: { dynamic: true } },
            });

            expect(mockExampleGenerator.generate).toHaveBeenCalled();
            expect(response.body).toBeDefined();

            const allExamples = flatMap(mockResource.responses, res =>
              flatMap(res.contents, content => content.examples),
            ).map(x => {
              if ('value' in x) return x.value;
            });

            allExamples.forEach(example => expect(response.body).not.toEqual(example));
            expect(response.body).toBe(generatedExample);
          });
        });
      });

      describe('and dynamic flag is false', () => {
        describe('and the example has been explicited', () => {
          it('should return the selected example', async () => {
            const response = await httpMocker.mock({
              input: mockInput,
              resource: mockResource,
              config: { mock: { dynamic: true, exampleKey: 'test key' } },
            });

            expect(response.body).toBeDefined();

            const selectedExample = flatMap(mockResource.responses, res =>
              flatMap(res.contents, content => content.examples),
            ).find(ex => ex.key === 'test key');

            expect(selectedExample).toBeDefined();
            expect(response.body).toEqual((selectedExample as INodeExample).value);
          });
        });
      });
    });
  });
});
