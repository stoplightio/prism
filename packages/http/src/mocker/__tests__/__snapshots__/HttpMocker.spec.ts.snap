// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HttpMocker mock() fails when called with no input 1`] = `"Http request is not defined"`;

exports[`HttpMocker mock() fails when called with no resource 1`] = `"Resource is not defined"`;

exports[`HttpMocker mock() when example is of type INodeExternalExample generates a dynamic example 1`] = `
Object {
  "body": "test value",
  "headers": Object {
    "Content-type": "test",
  },
  "statusCode": 202,
}
`;

exports[`HttpMocker mock() with invalid negotiator response returns static example 1`] = `
Object {
  "body": "hello",
  "headers": Object {
    "Content-type": "test",
  },
  "statusCode": 202,
}
`;

exports[`HttpMocker mock() with valid negotiator response returns dynamic example 1`] = `
Object {
  "body": "example value",
  "headers": Object {
    "Content-type": "test",
  },
  "statusCode": 202,
}
`;

exports[`HttpMocker mock() with valid negotiator response returns static example 1`] = `
Object {
  "body": "hello",
  "headers": Object {
    "Content-type": "test",
  },
  "statusCode": 202,
}
`;
